lib/block_view_lockups.go

```diff
@@ -0,0 +1,1784 @@
+	//       Performing the check here would be redundant and may lead to worse UX in the case of coins being
+	//       burned in the future making current lockups no longer an overflow. Checking here would also
+	//       create a DoS attack vector where a malicious entity takes out an extremely long-dated lockup
+	//       with the sole intent of saturating the CoinsInCirculationNanos field preventing others from locking up.
+
+	// For consolidation, we fetch equivalent LockedBalanceEntries.
+	lockedBalanceEntry := bav.GetLockedBalanceEntryForHODLerPKIDProfilePKIDExpirationTimestampNanoSecs(
+		hodlerPKID, profilePKID, txMeta.UnlockTimestampNanoSecs)
+	if lockedBalanceEntry == nil || lockedBalanceEntry.isDeleted {
+		lockedBalanceEntry = &LockedBalanceEntry{
+			HODLerPKID:                  hodlerPKID,
+			ProfilePKID:                 profilePKID,
+			ExpirationTimestampNanoSecs: txMeta.UnlockTimestampNanoSecs,
+			BalanceBaseUnits:            *uint256.NewInt(),
+		}
+	}
+	previousLockedBalanceEntry := *lockedBalanceEntry
+
+	// Check for overflow within the locked balance entry itself.
+	if uint256.NewInt().Sub(MaxUint256, yieldFromTxn).Lt(transactorBalanceNanos256) {
+		return 0, 0, nil,
+			errors.Wrap(RuleErrorCoinLockupYieldCausesOverflow, "_connectCoinLockup")
+	}
+	if profilePKID.IsZeroPKID() {
+		// Check if DeSo minted would overflow 2**64 in the transactor balance.
+		if uint256.NewInt().Sub(uint256.NewInt().SetUint64(math.MaxUint64), yieldFromTxn).Lt(transactorBalanceNanos256) {
+			return 0, 0, nil,
+				errors.Wrap(RuleErrorCoinLockupYieldCausesOverflow, "_connectCoinLockup")
+		}
+	}
+
+	// Increment the lockedBalanceEntry and update the view.
+	lockedBalanceEntry.BalanceBaseUnits = *uint256.NewInt().Add(&lockedBalanceEntry.BalanceBaseUnits, &lockupValue)
+	bav._setLockedBalanceEntry(lockedBalanceEntry)
+
+	// Add a UtxoOperation for easy reversion during disconnect.
+	utxoOpsForTxn = append(utxoOpsForTxn, &UtxoOperation{
+		Type:                       OperationTypeCoinLockup,
+		PrevTransactorBalanceEntry: prevTransactorBalanceEntry,
+		PrevLockedBalanceEntry:     &previousLockedBalanceEntry,
+	})
+
+	// Construct UtxoOps in the event this transaction is reverted.
+	return 0, 0, utxoOpsForTxn, nil
+}
+
+func CalculateLockupYield(
+	principal *uint256.Int,
+	apyYieldBasisPoints *uint256.Int,
+	durationNanoSecs *uint256.Int,
+) (*uint256.Int, error) {
+	// Note: We could compute either simple of compounding interest. While compounding interest is ideal from an
+	//       application perspective, it becomes incredibly difficult to implement from a numerical perspective.
+	//       This is because compound interest requires fractional exponents rather for computing the yield.
+	//       Determining overflow and preventing excessive money-printers becomes tricky in the compound interest case.
+	//       For this reason, we opt to use simple interest.
+	//
+	// Simple interest formula:
```

Comment:
I think Sofonias figured out compound interest for PoS. I think it's also fine to ship this the way it is though. It's arguably easier to reason about this way because you just get X% for locking up for Y duration, similar to how bonds work.